unittests/AST/ASTImporterTest.cpp

```diff
@@ -2184,38 +2182,128 @@ TEST_F(ImportFriendFunctions, LookupWithProto) {
+  To0 = FirstDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
   auto To1 = LastDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
   EXPECT_TRUE(To0->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
   EXPECT_TRUE(!To0->isInIdentifierNamespace(Decl::IDNS_Ordinary));
   EXPECT_TRUE(!To1->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
   EXPECT_TRUE(To1->isInIdentifierNamespace(Decl::IDNS_Ordinary));
 }
 
+TEST_F(ImportFriendFunctions, LookupWithProtoFirst) {
+  auto Pattern = functionDecl(hasName("f"));
+
+  Decl *FromTU =
+      getTuDecl("void f();"
+                "struct X { friend void f(); };",
+                Lang_CXX, "input0.cc");
+  auto From0 = FirstDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
+  auto From1 = LastDeclMatcher<FunctionDecl>().match(FromTU, Pattern);
+  ASSERT_TRUE(!From0->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
+  ASSERT_TRUE(From0->isInIdentifierNamespace(Decl::IDNS_Ordinary));
+  ASSERT_TRUE(From1->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
+  ASSERT_TRUE(From1->isInIdentifierNamespace(Decl::IDNS_Ordinary));
+  auto FromName = From0->getDeclName();
+  {
+    CXXRecordDecl *Class =
+        FirstDeclMatcher<CXXRecordDecl>().match(FromTU, cxxRecordDecl());
+    auto lookup_res = Class->noload_lookup(FromName);
+    ASSERT_EQ(lookup_res.size(), 0u);
+    lookup_res = cast<TranslationUnitDecl>(FromTU)->noload_lookup(FromName);
+    ASSERT_EQ(lookup_res.size(), 1u);
+  }
+
+  auto To0 = cast<FunctionDecl>(Import(From0, Lang_CXX));
+  auto ToName = To0->getDeclName();
+  Decl *ToTU = ToAST->getASTContext().getTranslationUnitDecl();
+
+  {
+    CXXRecordDecl *Class =
+        FirstDeclMatcher<CXXRecordDecl>().match(ToTU, cxxRecordDecl());
+    auto lookup_res = Class->noload_lookup(ToName);
+    EXPECT_EQ(lookup_res.size(), 0u);
+    lookup_res = cast<TranslationUnitDecl>(ToTU)->noload_lookup(ToName);
+    EXPECT_EQ(lookup_res.size(), 1u);
+  }
+
+  ASSERT_EQ(DeclCounter<FunctionDecl>().match(ToTU, Pattern), 2u);
+  To0 = FirstDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
+  auto To1 = LastDeclMatcher<FunctionDecl>().match(ToTU, Pattern);
+  EXPECT_TRUE(!To0->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
+  EXPECT_TRUE(To0->isInIdentifierNamespace(Decl::IDNS_Ordinary));
+  EXPECT_TRUE(To1->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
+  EXPECT_TRUE(To1->isInIdentifierNamespace(Decl::IDNS_Ordinary));
+}
+
+TEST_F(ImportFriendFunctions, ImportFriendChangesLookup) {
+  auto Pattern = functionDecl(hasName("f"));
+
+  Decl *FromTU0 = getTuDecl("void f();", Lang_CXX, "input0.cc");
+  FunctionDecl *FromD0 =
+      FirstDeclMatcher<FunctionDecl>().match(FromTU0, Pattern);
+  FunctionDecl *ToD0 = cast<FunctionDecl>(Import(FromD0, Lang_CXX));
+  auto FromName = FromD0->getDeclName();
+  auto ToName = ToD0->getDeclName();
+  Decl *ToTU = ToAST->getASTContext().getTranslationUnitDecl();
+  
+  EXPECT_TRUE(!FromD0->isInIdentifierNamespace(Decl::IDNS_OrdinaryFriend));
```

Comment:
All expectations on the "from" context should be `ASSERT_*` since if one of them is false then there is no point to continue the test case.